File: src/main/java/com/nerdearla/workshop/user/seller/client/SellerClient.java
package com.nerdearla.workshop.user.seller.client;

import com.nerdearla.workshop.user.seller.Seller;
import com.nerdearla.workshop.user.seller.error.SellerRetrievingError;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Component
public class SellerClient {

    private static final Logger LOGGER = LoggerFactory.getLogger(SellerClient.class);
    private final WebClient webClient;

    public SellerClient(WebClient webClient) {
        this.webClient = webClient;
    }

    public Seller getById(String id) {
        final Seller seller = webClient
                .get()
                .uri("/sellers/{sellerId}", id)
                .retrieve()
                .onStatus(HttpStatus::isError, this::handleError)
                .bodyToMono(Seller.class)
                .block();
        LOGGER.info("seller found: {}", seller);
        return seller;
    }

    private Mono<Throwable> handleError(ClientResponse response) {
        LOGGER.error("Error while communicating with seller service, {}", response);
        return Mono.error(new SellerRetrievingError());
    }
}

File: src/main/java/com/nerdearla/workshop/user/seller/client/SellerClient.kt
package com.nerdearla.workshop.user.seller.client

import com.nerdearla.workshop.shared.client.get
import com.nerdearla.workshop.shared.utils.CompanionLogger
import com.nerdearla.workshop.user.seller.Seller
import com.nerdearla.workshop.user.seller.error.SellerRetrievingError
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.ClientResponse
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

@Component
class SellerClient(
    private val client: WebClient
) {

    fun getBy(id: String): Seller =
        client.get<Seller>("/sellers/{sellerId}", id, handler = ::handleError)
            .log { info("seller found: {}", it) }

    private fun handleError(response: ClientResponse): Mono<Throwable> =
        Mono.error<Throwable>(SellerRetrievingError())
            .log { error("Error while communicating with seller service, {}", response) }

    companion object : CompanionLogger()
}
